taichi/com_mgr/kernel_compilation_manager.cpp

```diff
@@ -0,0 +1,217 @@
+  if (path_exists(filepath)) {
+    if (lock_with_file(lock_path)) {
+      auto _ = make_unlocker(lock_path);
+      offline_cache::load_metadata_with_checking(cached_data_, filepath);
+    } else {
+      TI_WARN(
+          "Lock {} failed. Please run 'ti cache clean -p {}' and try again.",
+          lock_path, config_.offline_cache_path);
+    }
+  }
+}
+
+const CompiledKernelData &KernelCompilationManager::load_or_compile(
+    const CompileConfig &compile_config,
+    const DeviceCapabilityConfig &caps,
+    const Kernel &kernel_def) {
+  auto cache_mode = get_cache_mode(compile_config, kernel_def);
+  const auto kernel_key = make_kernel_key(compile_config, caps, kernel_def);
+  auto cached_kernel = try_load_cached_kernel(kernel_def, kernel_key,
+                                              compile_config.arch, cache_mode);
+  return cached_kernel ? *cached_kernel
+                       : compile_and_cache_kernel(kernel_key, compile_config,
+                                                  caps, kernel_def);
+}
+
+void KernelCompilationManager::dump() {
+  if (caching_kernels_.empty()) {
+    return;
+  }
+
+  taichi::create_directories(config_.offline_cache_path);
+  auto filepath = join_path(config_.offline_cache_path, kMetadataFilename);
+  auto lock_path = join_path(config_.offline_cache_path, kMetadataLockName);
+
+  if (!lock_with_file(lock_path)) {
+    TI_WARN("Lock {} failed. Please run 'ti cache clean -p {}' and try again.",
+            lock_path, config_.offline_cache_path);
+    return;
+  }
+
+  auto _ = make_unlocker(lock_path);
+  CacheData data;
+  data.version[0] = TI_VERSION_MAJOR;
+  data.version[1] = TI_VERSION_MINOR;
+  data.version[2] = TI_VERSION_PATCH;
+  auto &kernels = data.kernels;
+  // Load old cached data
+  offline_cache::load_metadata_with_checking(data, filepath);
+  // Update the cached data
+  for (const auto *e : updated_data_) {
+    auto iter = kernels.find(e->kernel_key);
+    if (iter != kernels.end()) {
+      iter->second.last_used_at = e->last_used_at;
+    }
+  }
+  // Add new data
+  for (auto &[kernel_key, kernel] : caching_kernels_) {
+    if (kernel.cache_mode == CacheData::MemAndDiskCache) {
+      auto [iter, ok] = kernels.insert({kernel_key, std::move(kernel)});
+      data.size += ok ? iter->second.size : 0;
+    }
+  }
+  // Dump cached CompiledKernelData to disk
+  for (const auto &[_, k] : kernels) {
+    if (k.compiled_kernel_data) {
+      const auto arch = k.compiled_kernel_data->arch();
+      auto cache_filename = make_filename(k.kernel_key, arch);
+      if (try_lock_with_file(cache_filename)) {
+        std::ofstream fs{cache_filename, std::ios::out | std::ios::binary};
+        TI_ASSERT(fs.is_open());
+        k.compiled_kernel_data->dump(fs);
+      }
+    }
+  }
+  // Dump offline cache metadata
+  if (!kernels.empty()) {
+    write_to_binary_file(data, filepath);
+  }
+}
+
+void KernelCompilationManager::clean_offline_cache(
```

Comment:
does it break the cleaner test?

Reply:
I disable the offline cache cleaning test temporarily in this PR lol.

https://github.com/taichi-dev/taichi/blob/c107f72e47eb37c793e51825d0b92691361555c8/tests/python/test_offline_cache.py#L495-L496